include/zephyr/device.h

```diff
@@ -923,7 +926,694 @@ BUILD_ASSERT(sizeof(device_handle_t) == 2, "fix the linker scripts");
 	Z_INIT_ENTRY_DEFINE(DEVICE_NAME_GET(dev_name), init_fn,		\
 		(&DEVICE_NAME_GET(dev_name)), level, prio)
 
-#if CONFIG_HAS_DTS
+#ifdef CONFIG_LEGACY_DEVICE_MODEL
+
+/*
+ * The legacy device model can be used with the the new DEVICE_NEW_DEFINE() and
+ * DEVICE_DT_NEW_DEFINE() macros, which will expand to the legacy DEVICE_DEFINE()
+ * and DEVICE_DT_DEFINE() macros.
+ *
+ * This feature will allow for updating drivers to use the new multi API device
+ * model without breaking compatibility with the legacy device model, given that
+ * the device drivers only support a single API.
+ *
+ * Example from uart driver before update to new device model:
+ *
+ *     DEVICE_DT_DEFINE(node_id, ..., &uart_api_impl);
+ *
+ * Example from uart driver after update to new device model:
+ *
+ *     DEVICE_DT_NEW_DEFINE(node_id, ..., DEVICE_API(&uart_api_impl, uart));
+ *
+ * Using the legacy device model, both the legacy DEVICE_DT_DEFINE() and new
+ * driver DEVICE_DT_NEW_DEFINE() will expand identically. The device instances
+ * can be obtained using DEVICE_DT_GET(). The macros DEVICE_API_GET() and
+ * DEVICE_DT_API_GET() and their INST variants are not available. The new
+ * properties files and accompanying macros like DEVICE_DT_API_FOREACH(),
+ * DEVICE_DT_PROPERTY() etc. are available.
+ */
+
+#define DEVICE_NEW_DEFINE(dev_name, drv_name, init_fn, pm_dev_ptr, data_ptr,        \
```

Comment:
I think it's wrong to use `NEW` in APIs that are expected to replace the "old" ones.

Reply:
Do you have any suggestions?